// @effect/io/mjs/internal/fiberScope.mjs
var _a, _b;
import { globalValue } from "@effect/data/GlobalValue";
import * as FiberId from "@effect/io/FiberId";
import * as FiberMessage from "@effect/io/internal/fiberMessage";
/** @internal */
const FiberScopeSymbolKey = "@effect/io/FiberScope";
/** @internal */
export const FiberScopeTypeId = /*#__PURE__*/Symbol.for(FiberScopeSymbolKey);
/** @internal */
class Global {
  constructor() {
    this[_a] = FiberScopeTypeId;
    this.fiberId = FiberId.none;
    this.roots = new Set();
  }
  add(_runtimeFlags, child) {
    this.roots.add(child);
    child.addObserver(() => {
      this.roots.delete(child);
    });
  }
}
_a = FiberScopeTypeId;
/** @internal */
class Local {
  constructor(fiberId, parent) {
    this.fiberId = fiberId;
    this.parent = parent;
    this[_b] = FiberScopeTypeId;
  }
  add(_runtimeFlags, child) {
    this.parent.tell(FiberMessage.stateful(parentFiber => {
      parentFiber.addChild(child);
      child.addObserver(() => {
        parentFiber.removeChild(child);
      });
    }));
  }
}
_b = FiberScopeTypeId;
/** @internal */
export const unsafeMake = fiber => {
  return new Local(fiber.id(), fiber);
};
/** @internal */
export const globalScope = /*#__PURE__*/globalValue( /*#__PURE__*/Symbol.for("@effect/io/FiberScope/Global"), () => new Global());
