// @effect/io/mjs/Console.mjs
import * as internal from "@effect/io/internal/console";
import * as defaultConsole from "@effect/io/internal/defaultServices/console";
/**
 * @since 1.0.0
 * @category type ids
 */
export const TypeId = defaultConsole.TypeId;
/**
 * @since 1.0.0
 * @category default services
 */
export const withConsole = internal.withConsole;
/**
 * @since 1.0.0
 * @category default services
 */
export const setConsole = internal.setConsole;
/**
 * @since 1.0.0
 * @category accessor
 */
export const consoleWith = internal.consoleWith;
/**
 * @since 1.0.0
 * @category accessor
 */
export const assert = internal.assert;
/**
 * @since 1.0.0
 * @category accessor
 */
export const clear = internal.clear;
/**
 * @since 1.0.0
 * @category accessor
 */
export const count = internal.count;
/**
 * @since 1.0.0
 * @category accessor
 */
export const countReset = internal.countReset;
/**
 * @since 1.0.0
 * @category accessor
 */
export const debug = internal.debug;
/**
 * @since 1.0.0
 * @category accessor
 */
export const dir = internal.dir;
/**
 * @since 1.0.0
 * @category accessor
 */
export const dirxml = internal.dirxml;
/**
 * @since 1.0.0
 * @category accessor
 */
export const error = internal.error;
/**
 * @since 1.0.0
 * @category accessor
 */
export const group = internal.group;
/**
 * @since 1.0.0
 * @category accessor
 */
export const info = internal.info;
/**
 * @since 1.0.0
 * @category accessor
 */
export const log = internal.log;
/**
 * @since 1.0.0
 * @category accessor
 */
export const table = internal.table;
/**
 * @since 1.0.0
 * @category accessor
 */
export const time = internal.time;
/**
 * @since 1.0.0
 * @category accessor
 */
export const timeLog = internal.timeLog;
/**
 * @since 1.0.0
 * @category accessor
 */
export const trace = internal.trace;
/**
 * @since 1.0.0
 * @category accessor
 */
export const warn = internal.warn;
/**
 * @since 1.0.0
 * @category accessor
 */
export const withGroup = internal.withGroup;
/**
 * @since 1.0.0
 * @category accessor
 */
export const withTime = internal.withTime;
